### PR TITLE
[CLASSIC] Specs update for talent swapping

### PR DESCRIPTION
**Pull request by jazminite**
### Description

Lookup Classic spec by talent tree IF the icon is generic. Generic icons occur when the player changes talents during raid.

### Motivation

People are changing specs way more often with the new raid (Ulduar) release.

### Testing

- Test report URL: `/report/NdPnZVh9Bbk3jyHc/32-Normal+Auriaya+-+Kill+(3:37)`
- Screenshot(s):

[Live link before change](https://wowanalyzer.com/report/NdPnZVh9Bbk3jyHc/32-Normal+Auriaya+-+Kill+(3:37)):
image

After change:
image

### CODE REVIEW

File: src/game/SPECS.ts [hunk 1]
```diff
@@ -22,6 +22,7 @@ export interface Spec {
     spec: number;
   };
   icon?: string; // Classic only due to specId being 0
+  tree?: number; // Classic only for multiple spec users
 }
```
Comment (emallson) [inline src/game/SPECS.ts:46]: at this point, we should probably restructure this type a little:

```suggestion
interface BaseSpec {
  id: number;
  type?: string;
  index: number;
  className: string;
  specName?: string;
  role: number;
  primaryStat: string;
  ranking: { class: number; spec: number; };
}

interface RetailSpec extends BaseSpec {
  masterySpellId: number;
  masteryCoefficient: number;
}

interface ClassicSpec extends BaseSpec {
  icon: string;
  treeIndex: number;  // which tree (0,1,2) is for this spec. used as a fallback to try to do SOMETHING
}

type Spec = RetailSpec | ClassicSpec;
```

**Review by emallson — CHANGES_REQUESTED** [review]

File: src/game/SPECS.ts (hunk 1 above)
Comment (emallson) [reply src/game/SPECS.ts:46]: suggestion is broken, it didn't like suggesting things for unmodified text. would have to apply it manually but you get the idea

Comment (jazminite) [reply src/game/SPECS.ts:46]: I like the suggestion. I implemented it but it creates issues elsewhere (see below). Any suggestions?

```
Compiled with problems:X

ERROR in src/analysis/retail/druid/restoration/modules/core/Mastery.ts:284:40

TS2339: Property 'masteryCoefficient' does not exist on type 'Spec'.
  Property 'masteryCoefficient' does not exist on type 'ClassicSpec'.
    282 |           this.statTracker.currentMasteryRating,
    283 |           this.statTracker.statBaselineRatingPerPercent[STAT.MASTERY],
  > 284 |           this.selectedCombatant.spec?.masteryCoefficient,
        |                                        ^^^^^^^^^^^^^^^^^^
    285 |         );
    286 |       return buffBonus / healMasteryMult;
    287 |     };


ERROR in src/analysis/retail/monk/brewmaster/modules/talents/WeaponsOfOrder.tsx:30:68

TS2339: Property 'masteryCoefficient' does not exist on type 'Spec'.
  Property 'masteryCoefficient' does not exist on type 'ClassicSpec'.
    28 |
    29 |   get masteryBuffPercentage() {
  > 30 |     return BASE_MASTERY_PERCENTAGE * (this.selectedCombatant.spec?.masteryCoefficient || 1);
       |                                                                    ^^^^^^^^^^^^^^^^^^
    31 |   }
    32 |
    33 |   protected _resetCooldown(event: CastEvent) {


ERROR in src/parser/getConfig.ts:23:69

TS2339: Property 'icon' does not exist on type 'Spec'.
  Property 'icon' does not exist on type 'RetailSpec'.
    21 |     if (!config && player.icon) {
    22 |       config = expansionConfigs.find(
  > 23 |         (config) => config.spec.type === player.type && config.spec.icon === player.icon,
       |                                                                     ^^^^
    24 |       );
    25 |     }
    26 |   }


ERROR in src/parser/getConfig.ts:33:69

TS2339: Property 'tree' does not exist on type 'Spec'.
  Property 'tree' does not exist on type 'RetailSpec'.
    31 |       const maxTalent = talents.indexOf(Math.max(...talents));
    32 |       config = expansionConfigs.find(
  > 33 |         (config) => config.spec.type === player.type && config.spec.tree === maxTalent,
       |                                                                     ^^^^
    34 |       );
    35 |     }
    36 |   }


ERROR in src/parser/shared/modules/StatTracker.ts:528:64

TS2339: Property 'masteryCoefficient' does not exist on type 'Spec'.
  Property 'masteryCoefficient' does not exist on type 'ClassicSpec'.
    526 |   get baseMasteryPercentage() {
    527 |     const spellPoints = 8; // Spellpoint is a unit of mastery, each class has 8 base Spellpoints
  > 528 |     let mastery = (spellPoints * (this.selectedCombatant.spec?.masteryCoefficient || 1)) / 100;
        |                                                                ^^^^^^^^^^^^^^^^^^
    529 |     if (this.selectedCombatant.race === RACES.Dracthyr) {
    530 |       mastery += 0.018;
    531 |     }


ERROR in src/parser/shared/modules/StatTracker.ts:536:70

TS2339: Property 'masteryCoefficient' does not exist on type 'Spec'.
  Property 'masteryCoefficient' does not exist on type 'ClassicSpec'.
    534 |
    535 |   get hasMasteryCoefficient() {
  > 536 |     if (!this.selectedCombatant.spec || !this.selectedCombatant.spec.masteryCoefficient) {
        |                                                                      ^^^^^^^^^^^^^^^^^^
    537 |       return null;
    538 |     }
    539 |     return this.selectedCombatant.spec.masteryCoefficient;


ERROR in src/parser/shared/modules/StatTracker.ts:539:40

TS2339: Property 'masteryCoefficient' does not exist on type 'Spec'.
  Property 'masteryCoefficient' does not exist on type 'ClassicSpec'.
    537 |       return null;
    538 |     }
  > 539 |     return this.selectedCombatant.spec.masteryCoefficient;
        |                                        ^^^^^^^^^^^^^^^^^^
    540 |   }
    541 |
    542 |   get baseVersatilityPercentage() {


ERROR in src/parser/shared/modules/StatTracker.ts:667:38

TS2339: Property 'masteryCoefficient' does not exist on type 'Spec'.
  Property 'masteryCoefficient' does not exist on type 'ClassicSpec'.
    665 |         false,
    666 |         true,
  > 667 |         this.selectedCombatant.spec?.masteryCoefficient || 1,
        |                                      ^^^^^^^^^^^^^^^^^^
    668 |       )
    669 |     );
    670 |   }
```

Comment (emallson) [reply src/game/SPECS.ts:46]: we need some `isClassicSpec` / `isRetailSpec` helpers. i can pitch in on that tomorrow

**Review by jazminite — COMMENTED** [review]
lgtm :)

**Pull request merged**